### PR TITLE
Add @MainActor to waitForExpectations(timeout:handler:) #428

### LINKED ISSUE #428: Add `@MainActor` to `waitForExpectations(timeout:handler:)`
The Darwin implementation of XCTest added [`@MainActor`](https://developer.apple.com/documentation/swift/mainactor) to [`waitForExpectations(timeout:handler:)`](https://developer.apple.com/documentation/xctest/xctestcase/1500748-waitforexpectations). The open-source implementation should do the same.

### PR DESCRIPTION
**Pull request by grynspan**
This change makes `waitForExpectations()` as main-actor-isolated, as it is in Xcode's version of XCTest. It is marked `@preconcurrency @MainActor` so that existing synchronous functions can continue to use it without `await` or `@MainActor`.

Resolves #428.

### CODE REVIEW

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test

**Review by dennisweissmann — APPROVED** [review]

**Comment by grynspan** [comment]
@swift-ci please test Windows

**Comment by grynspan** [comment]
@swift-ci please clean test Windows

**Comment by grynspan** [comment]
@swift-ci please test

**Review by compnerd — APPROVED** [review]

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please clean test Windows

**Review by briancroom — APPROVED** [review]

File: Sources/XCTest/Public/Asynchronous/XCTestCase+Asynchronous.swift [hunk 1]
```diff
@@ -41,7 +41,7 @@ public extension XCTestCase {
     ///   these environments. To ensure compatibility of tests between
     ///   swift-corelibs-xctest and Apple XCTest, it is not recommended to pass
     ///   explicit values for `file` and `line`.
-    // FIXME: This should have `@MainActor` to match Xcode XCTest, but adding it causes errors in tests authored pre-Swift Concurrency which don't typically have `@MainActor`.
+    @preconcurrency @MainActor
     func waitForExpectations(timeout: TimeInterval, file: StaticString = #file, line: Int = #line, handler: XCWaitCompletionHandler? = nil) {
         precondition(Thread.isMainThread, "\(#function) must be called on the main thread")
         if currentWaiter != nil {
```
Comment (stmontgomery) [inline Sources/XCTest/Public/Asynchronous/XCTestCase+Asynchronous.swift:47]: This function is the only thing which uses the internal `currentWaiter` property on `XCTestCase`, and when I last touched this (in #340), I left a `// FIXME` comment on that property indicating that when we added `@MainActor` here, we could also consider adding `@MainActor` to that property too.

Now that you're doing that in this PR, could we also either add `@MainActor` to `currentWaiter`, or else delete the FIXME comment?

File: XCTest.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>IDEDidComputeMac32BitWarning</key>
+	<true/>
```
Comment (stmontgomery) [inline XCTest.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:6]: Can we omit this unrelated file Xcode added?

Comment (grynspan) [reply XCTest.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:6]: Oof, yes, thanks Xcode.

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows

File: Tests/Functional/Asynchronous/Expectations/main.swift [hunk 3]
```diff
@@ -551,7 +551,37 @@ class ExpectationsTestCase: XCTestCase {
         RunLoop.main.run(until: Date() + 1)
     }
 
-    static var allTests = {
+// CHECK: Test Case 'ExpectationsTestCase.test_waitForExpectationsFromMainActor' started at \d+-\d+-\d+ \d+:\d+:\d+\.\d+
+// CHECK: Test Case 'ExpectationsTestCase.test_waitForExpectationsFromMainActor' failed \(\d+\.\d+ seconds\)
+    func test_waitForExpectationsAsync() async {
```
Comment (stmontgomery) [inline Tests/Functional/Asynchronous/Expectations/main.swift:556]: I think the test function name in these `// CHECK` lines doesn't match the actual method name, which will likely cause CI to fail

Comment (grynspan) [reply Tests/Functional/Asynchronous/Expectations/main.swift:556]: Good eye, thanks.

File: Tests/Functional/Asynchronous/Expectations/main.swift [hunk 4]
```diff
@@ -603,15 +633,20 @@ class ExpectationsTestCase: XCTestCase {
             ("test_expectationCreationOnSecondaryThread", test_expectationCreationOnSecondaryThread),
             ("test_expectationCreationWhileWaiting", test_expectationCreationWhileWaiting),
             ("test_runLoopInsideDispatch", test_runLoopInsideDispatch),
+
+            // waitForExpectations() + @MainActor
+            ("test_waitForExpectationsAsync", asyncTest(test_waitForExpectationsAsync)),
+            ("test_waitForExpectationsFromMainActor", asyncTest { test_waitForExpectationsFromMainActor($0) }),
```
Comment (stmontgomery) [inline Tests/Functional/Asynchronous/Expectations/main.swift:639]: I wouldn't expect this one to use the `asyncTest` wrapper, since that means the wrapped test function will be invoked as if it were `async` and do the XCTestExpectation+XCTWaiter dance. Does plain
```suggestion
            ("test_waitForExpectationsFromMainActor", test_waitForExpectationsFromMainActor),
```
not work? If it doesn't, that feels like something else could be wrong

Comment (grynspan) [reply Tests/Functional/Asynchronous/Expectations/main.swift:639]: It's causing a compiler crash which I've filed internally as a radar.

Comment (grynspan) [reply Tests/Functional/Asynchronous/Expectations/main.swift:639]: rdar://111340003

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows

File: Tests/Functional/Asynchronous/Expectations/main.swift [hunk 5]
```diff
@@ -603,15 +633,20 @@ class ExpectationsTestCase: XCTestCase {
...
+            ("test_waitForExpectationsFromMainActor_async", asyncTest { test_waitForExpectationsFromMainActor_async($0) }),
```
Comment (stmontgomery) [inline Tests/Functional/Asynchronous/Expectations/main.swift:640]: Why can't this be
```suggestion
            ("test_waitForExpectationsFromMainActor_async", asyncTest(test_waitForExpectationsFromMainActor_async)),
```
Keep in mind that users never write `asyncTest(...)` or `asyncTest { ... }` themselves; SwiftPM inserts the former for them automatically if the test function has the `async` keyword

Comment (grynspan) [reply Tests/Functional/Asynchronous/Expectations/main.swift:640]: I already answered that. https://github.com/apple/swift-corelibs-xctest/pull/452#discussion_r1242597379

File: Tests/Functional/Asynchronous/Expectations/main.swift [hunk 6]
```diff
@@ -551,7 +551,37 @@ class ExpectationsTestCase: XCTestCase {
...
+    }
+
+// CHECK: Test Case 'ExpectationsTestCase.test_waitForExpectationsFromMainActor_async' started at \d+-\d+-\d+ \d+:\d+:\d+\.\d+
+// CHECK: Test Case 'ExpectationsTestCase.test_waitForExpectationsFromMainActor_async' failed \(\d+\.\d+ seconds\)
+    @MainActor func test_waitForExpectationsFromMainActor_async() async {
+      // Basic check that waitForExpectations() is functional and does not need
+      // the await keyword when used from a main-actor-isolated test function.
+      let expectation = self.expectation(description: "foo")
+      expectation.fulfill()
+      self.waitForExpectations(timeout: 0.0)
```
Comment (stmontgomery) [inline Tests/Functional/Asynchronous/Expectations/main.swift:581]: nit: these new tests appear to be using 2-space indent

Comment (grynspan) [reply Tests/Functional/Asynchronous/Expectations/main.swift]: Fixed.

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please clean test Windows

**Comment by grynspan** [comment]
Compiler crash happened again, so I'm just going to comment out the new _pro forma_ unit tests for now.

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows

**Comment by grynspan** [comment]
@swift-ci please test Windows

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows

**Comment by grynspan** [comment]
@swift-ci please test Windows

**Comment by grynspan** [comment]
@swift-ci please test

**Comment by grynspan** [comment]
@swift-ci please test Windows
